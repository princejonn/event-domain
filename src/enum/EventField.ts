export enum EventField {
  CAUSATION_ID = "ca",
  CORRELATION_ID = "co",
  DATA = "da",
  ID = "id",
  NAME = "na",
  TIMESTAMP = "tc",
}
