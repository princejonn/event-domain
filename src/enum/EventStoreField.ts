export enum EventStoreField {
  CAUSATION_ID = "ca",
  EVENTS = "ev",
  LOAD_EVENTS = "lo",
  PATH = "pa",
  REVISION = "re",
  TIMESTAMP = "tc",
}
