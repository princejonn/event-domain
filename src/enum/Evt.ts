export enum Evt {
  EVENT = "event",
  ERROR = "error",
}
