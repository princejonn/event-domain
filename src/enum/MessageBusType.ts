export enum MessageBusType {
  AMQP = "amqp",
  CUSTOM = "custom",
  MEMORY = "memory",
}
