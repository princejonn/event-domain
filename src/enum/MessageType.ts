export enum MessageType {
  COMMAND = "command",
  EVENT = "event",
  TIMEOUT = "timeout",
  UNKNOWN = "unknown",
}
