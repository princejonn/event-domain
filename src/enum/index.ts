export * from "./EventField";
export * from "./EventStoreField";
export * from "./Evt";
export * from "./MessageBusType";
export * from "./MessageType";
export * from "./MetaState";
export * from "./SagaStoreField";
export * from "./StoreType";
export * from "./ViewStoreField";
