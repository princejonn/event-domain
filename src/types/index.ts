export * from "./aggregate";
export * from "./aggregate-command-handler";
export * from "./aggregate-domain";
export * from "./aggregate-event-handler";
export * from "./app";
export * from "./error";
export * from "./event-emitter";
export * from "./event-store";
export * from "./generic";
export * from "./guid";
export * from "./handler";
export * from "./message";
export * from "./message-bus";
export * from "./mongo";
export * from "./saga";
export * from "./saga-domain";
export * from "./saga-event-handler";
export * from "./saga-store";
export * from "./scan";
export * from "./view";
export * from "./view-domain";
export * from "./view-event-handler";
export * from "./view-store";
