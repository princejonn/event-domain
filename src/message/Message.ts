import { AggregateIdentifier, IMessage, MessageOptions } from "../types";
import { MessageType } from "../enum";
import { isString } from "lodash";
import { parseISO } from "date-fns";
import { randomUUID } from "crypto";

export abstract class Message implements IMessage {
  readonly id: string;
  readonly name: string;
  readonly type: MessageType;

  readonly aggregate: AggregateIdentifier;
  readonly causationId: string;
  readonly correlationId: string;
  readonly data: Record<string, any>;
  readonly delay: number;
  readonly mandatory: boolean;
  readonly timestamp: Date;

  protected constructor(options: MessageOptions, causation?: IMessage) {
    this.id = options.id || randomUUID();
    this.name = options.name;
    this.type = options.type || MessageType.UNKNOWN;

    this.aggregate = options.aggregate;
    this.causationId = options.causationId || causation?.id || this.id;
    this.correlationId = options.correlationId || causation?.correlationId || randomUUID();
    this.data = options.data || {};
    this.delay = options.delay || 0;
    this.mandatory = options.mandatory || false;

    const timestamp = options.timestamp || new Date();
    this.timestamp = isString(timestamp) ? parseISO(timestamp) : timestamp;
  }
}
