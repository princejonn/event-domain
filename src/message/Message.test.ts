import MockDate from "mockdate";
import { IMessage, MessageOptions } from "../types";
import { Message } from "./Message";
import { MessageType } from "../enum";

MockDate.set("2022-01-01T08:00:00.000Z");

jest.mock("crypto", () => ({
  randomUUID: jest.fn(() => "e64b576a-022e-4dd4-8c97-32408012f8d1"),
}));

class TestMessage extends Message {
  constructor(options: MessageOptions, causation?: IMessage) {
    super(options, causation);
  }
}

describe("Message", () => {
  test("should create a message", () => {
    expect(
      new TestMessage({
        name: "name",
        aggregate: {
          id: "id",
          name: "name",
          context: "context",
        },
      }),
    ).toMatchSnapshot();
  });

  test("should set id as correlation by default", () => {
    expect(
      new TestMessage({
        id: "bacbc463-f3dc-4978-a719-8bcdddb54931",
        name: "name",
        aggregate: {
          id: "id",
          name: "name",
          context: "context",
        },
      }),
    ).toMatchSnapshot();
  });

  test("should respect options", () => {
    expect(
      new TestMessage({
        id: "8a7d01b6-dc4c-42a2-a444-4f4003747994",
        name: "name",
        type: MessageType.EVENT,
        aggregate: { id: "id", name: "name", context: "context" },
        causationId: "83863312-21d6-46a6-b34d-1fb660de5bea",
        correlationId: "97d2378f-43d1-4446-88b4-bcd4cb7aa44b",
        data: { value: 1 },
        delay: 999,
        mandatory: true,
        timestamp: new Date("1999-01-01"),
      }),
    ).toMatchSnapshot();
  });

  test("should prioritize options over causation", () => {
    expect(
      new TestMessage(
        {
          name: "name",
          aggregate: { id: "id", name: "name", context: "context" },
          causationId: "83863312-21d6-46a6-b34d-1fb660de5bea",
          correlationId: "97d2378f-43d1-4446-88b4-bcd4cb7aa44b",
        },
        new TestMessage({
          id: "9bdbf988-c405-4e07-bf43-e30a10c8ed9a",
          name: "name",
          aggregate: { id: "id", name: "name", context: "context" },
          correlationId: "dd514fa0-08eb-404e-993e-a9687977bd23",
        }),
      ),
    ).toMatchSnapshot();
  });

  test("should fall back to causation", () => {
    expect(
      new TestMessage(
        {
          name: "name",
          aggregate: { id: "id", name: "name", context: "context" },
        },
        new TestMessage({
          id: "9bdbf988-c405-4e07-bf43-e30a10c8ed9a",
          name: "name",
          aggregate: { id: "id", name: "name", context: "context" },
          correlationId: "dd514fa0-08eb-404e-993e-a9687977bd23",
        }),
      ),
    ).toMatchSnapshot();
  });

  test("should convert timestamp to Date when string", () => {
    expect(
      new TestMessage({
        name: "name",
        aggregate: { id: "id", name: "name", context: "context" },
        timestamp: "2000-10-10T09:00:00.888Z",
      }),
    ).toMatchSnapshot();
  });
});
