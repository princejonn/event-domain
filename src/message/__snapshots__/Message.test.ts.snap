// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message should convert timestamp to Date when string 1`] = `
TestMessage {
  "aggregate": Object {
    "context": "context",
    "id": "id",
    "name": "name",
  },
  "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "data": Object {},
  "delay": 0,
  "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "mandatory": false,
  "name": "name",
  "timestamp": 2000-10-10T09:00:00.888Z,
  "type": "unknown",
}
`;

exports[`Message should create a message 1`] = `
TestMessage {
  "aggregate": Object {
    "context": "context",
    "id": "id",
    "name": "name",
  },
  "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "data": Object {},
  "delay": 0,
  "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "mandatory": false,
  "name": "name",
  "timestamp": 2022-01-01T08:00:00.000Z,
  "type": "unknown",
}
`;

exports[`Message should fall back to causation 1`] = `
TestMessage {
  "aggregate": Object {
    "context": "context",
    "id": "id",
    "name": "name",
  },
  "causationId": "9bdbf988-c405-4e07-bf43-e30a10c8ed9a",
  "correlationId": "dd514fa0-08eb-404e-993e-a9687977bd23",
  "data": Object {},
  "delay": 0,
  "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "mandatory": false,
  "name": "name",
  "timestamp": 2022-01-01T08:00:00.000Z,
  "type": "unknown",
}
`;

exports[`Message should prioritize options over causation 1`] = `
TestMessage {
  "aggregate": Object {
    "context": "context",
    "id": "id",
    "name": "name",
  },
  "causationId": "83863312-21d6-46a6-b34d-1fb660de5bea",
  "correlationId": "97d2378f-43d1-4446-88b4-bcd4cb7aa44b",
  "data": Object {},
  "delay": 0,
  "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "mandatory": false,
  "name": "name",
  "timestamp": 2022-01-01T08:00:00.000Z,
  "type": "unknown",
}
`;

exports[`Message should respect options 1`] = `
TestMessage {
  "aggregate": Object {
    "context": "context",
    "id": "id",
    "name": "name",
  },
  "causationId": "83863312-21d6-46a6-b34d-1fb660de5bea",
  "correlationId": "97d2378f-43d1-4446-88b4-bcd4cb7aa44b",
  "data": Object {
    "value": 1,
  },
  "delay": 999,
  "id": "8a7d01b6-dc4c-42a2-a444-4f4003747994",
  "mandatory": true,
  "name": "name",
  "timestamp": 1999-01-01T00:00:00.000Z,
  "type": "event",
}
`;

exports[`Message should set id as correlation by default 1`] = `
TestMessage {
  "aggregate": Object {
    "context": "context",
    "id": "id",
    "name": "name",
  },
  "causationId": "bacbc463-f3dc-4978-a719-8bcdddb54931",
  "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "data": Object {},
  "delay": 0,
  "id": "bacbc463-f3dc-4978-a719-8bcdddb54931",
  "mandatory": false,
  "name": "name",
  "timestamp": 2022-01-01T08:00:00.000Z,
  "type": "unknown",
}
`;
